File: clients/Go/model/debug_command.go
package model

import "io"
import . "aicup2020/stream"

type DebugCommand interface {
    Write(writer io.Writer)
}
func ReadDebugCommand(reader io.Reader) DebugCommand {
    switch ReadInt32(reader) {
        case 0:
            return ReadDebugCommandAdd(reader)
        case 1:
            return ReadDebugCommandClear(reader)
        case 2:
            return ReadDebugCommandSetAutoFlush(reader)
        case 3:
            return ReadDebugCommandFlush(reader)
    }
    panic("Unexpected tag value")
}

type DebugCommandAdd struct {
    Data DebugData
}
func NewDebugCommandAdd(data DebugData) DebugCommandAdd {
    return DebugCommandAdd {
        Data: data,
    }
}
func ReadDebugCommandAdd(reader io.Reader) DebugCommandAdd {
    result := DebugCommandAdd {}
    result.Data = ReadDebugData(reader)
    return result
}
func (value DebugCommandAdd) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    value.Data.Write(writer)
}

type DebugCommandClear struct {
}
func NewDebugCommandClear() DebugCommandClear {
    return DebugCommandClear {
    }
}
func ReadDebugCommandClear(reader io.Reader) DebugCommandClear {
    result := DebugCommandClear {}
    return result
}
func (value DebugCommandClear) Write(writer io.Writer) {
    WriteInt32(writer, 1)
}

type DebugCommandSetAutoFlush struct {
    Enable bool
}
func NewDebugCommandSetAutoFlush(enable bool) DebugCommandSetAutoFlush {
    return DebugCommandSetAutoFlush {
        Enable: enable,
    }
}
func ReadDebugCommandSetAutoFlush(reader io.Reader) DebugCommandSetAutoFlush {
    result := DebugCommandSetAutoFlush {}
    result.Enable = ReadBool(reader)
    return result
}
func (value DebugCommandSetAutoFlush) Write(writer io.Writer) {
    WriteInt32(writer, 2)
    WriteBool(writer, value.Enable)
}

type DebugCommandFlush struct {
}
func NewDebugCommandFlush() DebugCommandFlush {
    return DebugCommandFlush {
    }
}
func ReadDebugCommandFlush(reader io.Reader) DebugCommandFlush {
    result := DebugCommandFlush {}
    return result
}
func (value DebugCommandFlush) Write(writer io.Writer) {
    WriteInt32(writer, 3)
}

File: clients/Python-PyPy/model/debug_command.py
class DebugCommand:
    @staticmethod
    def read_from(stream):
        tag = stream.read_int()
        if tag == Add.TAG:
            return DebugCommand.Add.read_from(stream)
        if tag == Clear.TAG:
            return DebugCommand.Clear.read_from(stream)
        if tag == SetAutoFlush.TAG:
            return DebugCommand.SetAutoFlush.read_from(stream)
        if tag == Flush.TAG:
            return DebugCommand.Flush.read_from(stream)
        raise Exception("Unexpected tag value")

from .debug_data import DebugData
class Add(DebugCommand):
    TAG = 0
    def __init__(self, data):
        self.data = data
    @staticmethod
    def read_from(stream):
        data = DebugData.read_from(stream)
        return Add(data)
    def write_to(self, stream):
        stream.write_int(self.TAG)
        self.data.write_to(stream)
    def __repr__(self):
        return "Add(" + \
            repr(self.data) + \
            ")"
DebugCommand.Add = Add
class Clear(DebugCommand):
    TAG = 1
    def __init__(self):
        pass
    @staticmethod
    def read_from(stream):
        return Clear()
    def write_to(self, stream):
        stream.write_int(self.TAG)
    def __repr__(self):
        return "Clear(" + \
            ")"
DebugCommand.Clear = Clear
class SetAutoFlush(DebugCommand):
    TAG = 2
    def __init__(self, enable):
        self.enable = enable
    @staticmethod
    def read_from(stream):
        enable = stream.read_bool()
        return SetAutoFlush(enable)
    def write_to(self, stream):
        stream.write_int(self.TAG)
        stream.write_bool(self.enable)
    def __repr__(self):
        return "SetAutoFlush(" + \
            repr(self.enable) + \
            ")"
DebugCommand.SetAutoFlush = SetAutoFlush
class Flush(DebugCommand):
    TAG = 3
    def __init__(self):
        pass
    @staticmethod
    def read_from(stream):
        return Flush()
    def write_to(self, stream):
        stream.write_int(self.TAG)
    def __repr__(self):
        return "Flush(" + \
            ")"
DebugCommand.Flush = Flush
